### PR TITLE
Add a Design Process Policy

### PR DESCRIPTION
**Pull request by mattcaswell**
There is currently no policy which describes how designs for
significant solutions to be incorporated into OpenSSL are arrived at.
This policy provides a process for doing that.

### CODE REVIEW

File: policies/design-process.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+OTC. In some cases more significant rework may be required, in which case the
+OTC may choose to step back to any previous step in this design process.
+
+In some cases the OTC may agree a solution which has certain details
+deliberately omitted in order to defer work on those details until a later stage
+of the implementation project. In such cases this process iterates back to the
+"Refine and document" stage to fill in those details when they are eventually
+required, and the subsequent amended design document must be agreed again by the
+OTC.
+
```
Comment (beldmit) [inline policies/design-process.md:126]: We also need some sort of a review solution. The original design could miss some crucial points or see a better approach, so the design will require some changes.

**Review by beldmit — CHANGES_REQUESTED** [review]

File: policies/design-process.md (hunk 1 above)
Comment (mattcaswell) [reply policies/design-process.md:126]: I saw this as happening on github in the pull request, i.e. in the "refine and document" step the design document is made available as a PR on which people can make comments. The document is updated accordingly to take account of comments. Eventually the document stabilises sufficiently that OTC are asked to agree it. Either they agree, or they think it needs more work in which case it goes back to "refine and document" as stated above, i.e. "The OTC may choose not to agree a solution. In this case the process should iterate back to the "Refine and document" step to incorporate feedback from the OTC."

File: policies/design-process.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+policy should be applied recursively to address the design for each sub-problem
+individually before incorporating all of the outputs into a single large
+design document.
+
+For such large problems a design workshop would normally be held to identify
+the sub-problems that require individual designs.
+
+### Identify candidate solutions
+
+Once the problem area is understood candidate solutions (one or more) that meet
```
Comment (slontis) [inline policies/design-process.md:64]: understood,

File: policies/design-process.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+progress and to understand the trajectory of the solution into the future. Some
+details may be omitted until a later stage of the development project if
+appropriate and they are not needed until later.
+
+The design document should be available for public comment in the form of a
+pull request.
+
+### Agree solution
+
+In this step the OTC agree the solution as it is described in the design
```
Comment (slontis) [inline policies/design-process.md:111]: to the solution?

File: policies/design-process.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+document as produced by the design team. Once agreed the implementation may
+proceed and the design document should be merged into the otc repository.
+
+The OTC may choose not to agree a solution. In this case the process should
+iterate back to the "Refine and document" step to incorporate feedback from the
+OTC. In some cases more significant rework may be required, in which case the
+OTC may choose to step back to any previous step in this design process.
+
+In some cases the OTC may agree a solution which has certain details
```
Comment (slontis) [inline policies/design-process.md:120]: to a

File: policies/design-process.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+deliberately omitted in order to defer work on those details until a later stage
```
Comment (slontis) [inline policies/design-process.md:121]: The assumptions of a solution should be listed (including omissions)

File: policies/design-process.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+Transparency and Community Involvement
+--------------------------------------
+
+It should be possible for community members to see the progress of designs
+through the process and to pass comment on the proposed design prior to its
+eventual agreement by OTC. As a minimum:
+
+* The draft design document containing the candidate solutions should be
+available as a pull request for public comment for at least one week prior to
+the selection of the solution. As an alternative community members may be
```
Comment (slontis) [inline policies/design-process.md:136]: alternative,

File: policies/design-process.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
+Policy for Creating Designs
+===========================
+
+From time to time the OTC may require that certain problems should create a
```
Comment (paulidale) [inline policies/design-process.md:4]: _...certain problems should have a design document..._

The problems aren't creating anything :)

File: policies/design-process.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+meeting minutes.
+
+Outputs
+-------
+
+The result of this process should be a design document in markdown format,
+reviewed via a pull request and approved for merge into the otc git repository.
+
+Other outputs may include feedback on the requirements and constraints that
+were the input into the process. For example some candidate solutions that are
```
Comment (paulidale) [inline policies/design-process.md:35]: comma after _some_?

File: policies/design-process.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+through the process and to pass comment on the proposed design prior to its
+eventual agreement by OTC. As a minimum:
+
+* The draft design document containing the candidate solutions should be
+available as a pull request for public comment for at least one week prior to
+the selection of the solution. As an alternative community members may be
+offered an open invite to join the solution selection meeting.
+
+* The eventual refined documented design should be available for public comment
+as a pull request for at least one week prior to eventual agreement of the
```
Comment (paulidale) [inline policies/design-process.md:140]: drop _eventual_
`s/of/as to/`

**Review by paulidale — COMMENTED** [review]
Looks pretty good.

File: policies/design-process.md (hunk 5 above)
Comment (mattcaswell) [reply policies/design-process.md:121]: I added a sentence about this to the "Refine and document" section

File: policies/design-process.md (hunk 8 above)
Comment (mattcaswell) [reply policies/design-process.md:35]: I assume you really meant comma after "For example". Added.

File: policies/design-process.md (hunk 9 above)
Comment (mattcaswell) [reply policies/design-process.md:140]: This change doesn't make sense to me?

File: policies/design-process.md (hunk 3 above)
Comment (mattcaswell) [reply policies/design-process.md:111]: To me "agree a solution" seems more correct than "agree to a solution"

File: policies/design-process.md (hunk 4 above)
Comment (mattcaswell) [reply policies/design-process.md:120]: As above

**Comment by mattcaswell** [comment]
I've pushed an update with the updates from the feedback above as well as incorporating feedback from the OTC meeting last week. Are there any more comments or can we start a vote on this?

File: policies/design-process.md [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+Inputs
+------
+
+Prior to commencing design work the OTC shall identify:
+
+1) The business requirements for the problem to be solved (obtained from the OMC)
+2) The scope of the specific problem area to be addressed
+3) Any additional design constraints that the OTC wishes to impose
+4) Any other constraints that may apply (e.g. timescales)
+5) The individual(s) that will work on developing the design (refered to as the
```
Comment (t8m) [inline policies/design-process.md:27]: Typo: referred

File: policies/design-process.md (hunk 3 above)
Comment (t8m) [reply policies/design-process.md:125]: Maybe `Approve solution` instead?

File: policies/design-process.md [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+to iterate over the candidate solutions again and either make refinements or
+identify additional options.
+
+### Refine and document
+
+In this step the design team provide further detail on the selected solution
+option. There should be sufficient detail to enable the implementation work to
+progress and to understand the trajectory of the solution into the future. Some
+details may be omitted until a later stage of the development project if
+appropriate and they are not needed until later. The assumptions of a solution
```
Comment (t8m) [inline policies/design-process.md:117]: maybe: `...if appropriate and such a level of detail is not needed until later development stages.`?

File: policies/design-process.md [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+Transparency and Community Involvement
+--------------------------------------
+
+It should be possible for community members to see the progress of designs
+through the process and to pass comment on the proposed design prior to its
+eventual agreement by OTC. As a minimum:
+
+* The draft design document containing the candidate solutions should be
+available as a pull request for public comment for at least one week prior to
+the selection of the solution. As an alternative, community members may be
```
Comment (t8m) [inline policies/design-process.md:150]: IMO the one week for the PR before the selection should be always kept. So perhaps: `In addition to that, community members may...`

**Review by t8m — CHANGES_REQUESTED** [review]

File: policies/design-process.md (hunk 3 above)
Comment (mattcaswell) [reply policies/design-process.md]: That works too. I changed it to that.

File: policies/design-process.md [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+candidate solutions, or it could be in the form of an open invite for community
+members to join the workshop where the solution is selected.
+
+Optionally the OTC could select no solution and instead direct the design team
+to iterate over the candidate solutions again and either make refinements or
+identify additional options.
+
+### Refine and document
+
+In this step the design team provide further detail on the selected solution
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply policies/design-process.md:113]: "provide" seems correct here.

**Comment by mattcaswell** [comment]
Further fixup pushed addressing additional feedback

File: policies/design-process.md [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+Outputs
+-------
+
+The result of this process should be a design document in markdown format,
+reviewed via a pull request and approved for merge into the main source git
+repository. The design document must include:
+
+ - Any assumptions that have been made
+ - Any omissions that may be filled in later
+ - Any APIs that the implementation will provide
```
Comment (kroeckx) [inline policies/design-process.md:42]: Does this mean that (minimal) documentation should written for the API as part of the design document?

File: policies/design-process.md [hunk 15]
```diff
@@ -0,0 +1,154 @@
+Policy for Creating Designs
+===========================
+
+From time to time the OTC may require that certain problems should have a
+design document created for them before implementation should start. In that
+case this policy will apply.
```
Comment (kroeckx) [inline policies/design-process.md:6]: It's unclear in which case the policy needs to applied. Will this be required for all new features? Or when we think the feature reaches some threshold of size, difficulty, ... ? Should 1 OTC member just ask for it? What if there is a disagreement that one is needed or not? Can non OTC members ask for it?

File: policies/design-process.md [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+
+Typically these will be identified during an OTC meeting and documented in the
+meeting minutes.
+
+Outputs
+-------
+
+The result of this process should be a design document in markdown format,
+reviewed via a pull request and approved for merge into the main source git
+repository. The design document must include:
```
Comment (kroeckx) [inline policies/design-process.md:38]: Does this mean the approval of a design follows our normal approval process for code changes? Or do we want the OTC to approve it with a vote?

File: policies/design-process.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (kaduk) [reply policies/design-process.md:113]: It seems to hinge on whether the noun "design team" is considered singular or plural.  I would say "the design team provides", but "the design team members provide".

File: policies/design-process.md (hunk 14 above)
Comment (mattcaswell) [reply policies/design-process.md:42]: I've added this sentence to the document to clarify this "This should include some minimal documentation of the APIs (not necessarily end-user focussed) to show how they should be used."

File: policies/design-process.md (hunk 15 above)
Comment (mattcaswell) [reply policies/design-process.md:6]: This only applies if an OTC *decision* occurs. Its not sufficient for 1 OTC member to ask for it. If consensus cannot be reached then a vote would be required. It's not for all new features and there are no formal thresholds defined - only those features where OTC collectively decides it is required. Possibly future amendments might add more rules in this area - but for now I think this is sufficient. I've clarified the first paragraph to say:

"From time to time the OTC may decide that certain problems should have a
design document created for them before implementation should start. Such a
decision should be minuted in OTC meeting minutes or recorded via a vote. In
that case this policy will apply."

File: policies/design-process.md (hunk 16 above)
Comment (mattcaswell) [reply policies/design-process.md:38]: No. Approval is via OTC decision. I've clarified this to say:

"The result of this process should be a design document in markdown format,
reviewed via a pull request and approved for merge into the main source git
repository via an OTC decision (minutes in OTC meeting minutes or recorded via
a vote)."

**Comment by mattcaswell** [comment]
I've updated this PR to take account of the latest feedback and to incorporate feedback from today's OTC meeting. Barring further significant comments I plan to open a vote on this tomorrow.

**Comment by t-j-h** [comment]
Vote [+1]

**Comment by mattcaswell** [comment]
@openssl/otc: VOTE NOW OPEN.

Please record your votes in this PR

**Comment by mattcaswell** [comment]
+1

**Comment by romen** [comment]
Vote [+1]

**Comment by t8m** [comment]
Vote [+1]

**Comment by mspncp** [comment]
Vote [+1]

**Comment by beldmit** [comment]
Vote [+1]

**Comment by kroeckx** [comment]
Vote [+1]

**Comment by paulidale** [comment]
Volte [+1]

**Comment by slontis** [comment]
Vote [+1]

**Comment by levitte** [comment]
vote [+1]

**Comment by t-j-h** [comment]
Closed - pass - 10 Yes, 0 No, 0 abstension.

**Pull request closed without merging**